File: pcc/py_runtime/src/py_print_fmt.c
/* pcc/py_runtime/src/py_print_fmt.c
 *
 * Formatter + high-level print entry points split out of py_print.c.
 * py_format / py_print / py_print_many stay here (heavy on stdio
 * format strings that pcc-Python cannot yet express). The pure
 * sys-stdout.write / sys-stderr.write path lives in py_print_sys.c
 * so it can be independently replaced by the pcc-Python port.
 */

#include "py_internal.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <inttypes.h>
#include <unistd.h>

extern int py_format_try_cpy_object_into_fd(int fd, void *obj, int32_t tag);
extern PyObject *py_exc_get_message(PyObject *exc);
extern PyObject *py_obj_str(PyObject *o);
extern PyObject *py_obj_repr(PyObject *o);
extern int64_t py_exc_matches(PyObject *exc, PyObject *type);

static void py_format(FILE *fp, PyObject *o);
static void py_format_repr(FILE *fp, PyObject *o);
static void py_format_bytes(FILE *fp, PyObject *o);

static PyObject *fmt_list_item(PyObject *owner, PyListObject *l, int64_t i) {
    return pcc_gc_load_ptr(owner, &l->items[i]);
}

static PyObject *fmt_tuple_item(PyObject *owner, PyTupleObject *t, int64_t i) {
    return pcc_gc_load_ptr(owner, &t->items[i]);
}

static void py_format_int(FILE *fp, PyObject *o) {
    if (PY_IS_TAGGED_INT(o)) {
        fprintf(fp, "%" PRId64, py_untag_int(o));
        return;
    }
    const PyIntObject *b = (const PyIntObject *)o;
    int overflow = 0;
    int64_t v = py_bigint_to_i64(b, &overflow);
    if (!overflow) {
        fprintf(fp, "%" PRId64, v);
        return;
    }
    char *s = py_bigint_to_cstr(b);
    if (s == NULL) return;
    fputs(s, fp);
    free(s);
}

static void py_format_float(FILE *fp, PyObject *o) {
    /* CPython shortest-round-trip repr via the shared helper (mirrors the
     * pcc-Python ports). The old path used "%g" (6 significant figures), which
     * truncated e.g. 10/3 to "3.33333". */
    PyObject *s = py_float_repr_shortest(o);
    if (s == NULL) {
        return;
    }
    fwrite(py_str_utf8(s), 1, (size_t)py_str_byte_len(s), fp);
    py_decref(s);
}

static void py_format_str(FILE *fp, PyObject *o) {
    PyStrObject *s = (PyStrObject *)o;
    if (s->byte_len <= 0) return;
    fwrite(s->data, 1, (size_t)s->byte_len, fp);
}

static void py_format_str_repr(FILE *fp, PyObject *o) {
    PyStrObject *s = (PyStrObject *)o;
    fputc('\'', fp);
    for (int64_t i = 0; i < s->byte_len; i++) {
        unsigned char c = (unsigned char)s->data[i];
        switch (c) {
            case '\\':
                fputs("\\\\", fp);
                break;
            case '\'':
                fputs("\\'", fp);
                break;
            case '\n':
                fputs("\\n", fp);
                break;
            case '\r':
                fputs("\\r", fp);
                break;
            case '\t':
                fputs("\\t", fp);
                break;
            default:
                if (c < 32 || c == 127) {
                    fprintf(fp, "\\x%02x", (unsigned)c);
                } else {
                    fputc((int)c, fp);
                }
                break;
        }
    }
    fputc('\'', fp);
}

static void py_format_bytes(FILE *fp, PyObject *o) {
    PyBytesObject *b = (PyBytesObject *)o;
    fputs("b'", fp);
    for (int64_t i = 0; i < b->byte_len; i++) {
        unsigned char c = (unsigned char)b->data[i];
        switch (c) {
            case '\\':
                fputs("\\\\", fp);
                break;
            case '\'':
                fputs("\\'", fp);
                break;
            case '\n':
                fputs("\\n", fp);
                break;
            case '\r':
                fputs("\\r", fp);
                break;
            case '\t':
                fputs("\\t", fp);
                break;
            default:
                /* bytes repr shows only printable ASCII (32..126) raw;
                 * control bytes (<32), DEL (127) and all high bytes (>=128)
                 * escape as \xNN. The old ``c == 127`` missed 128..255, so
                 * b'\xcf\x80' printed the raw UTF-8 instead of the escapes. */
                if (c < 32 || c >= 127) {
                    fprintf(fp, "\\x%02x", (unsigned)c);
                } else {
                    fputc((int)c, fp);
                }
                break;
        }
    }
    fputc('\'', fp);
}

/* bytearray repr: ``bytearray(b'...')`` — the byte escaping is identical to
 * bytes (same layout: byte_len + data), wrapped in ``bytearray( ... )``. */
static void py_format_bytearray(FILE *fp, PyObject *o) {
    fputs("bytearray(", fp);
    py_format_bytes(fp, o);
    fputc(')', fp);
}

static void py_format_list(FILE *fp, PyObject *o) {
    PyListObject *l = (PyListObject *)o;
    fputc('[', fp);
    for (int64_t i = 0; i < l->length; i++) {
        if (i > 0) fputs(", ", fp);
        py_format_repr(fp, fmt_list_item(o, l, i));
    }
    fputc(']', fp);
}

static void py_format_tuple(FILE *fp, PyObject *o) {
    PyTupleObject *t = (PyTupleObject *)o;
    fputc('(', fp);
    for (int64_t i = 0; i < t->len; i++) {
        if (i > 0) fputs(", ", fp);
        py_format_repr(fp, fmt_tuple_item(o, t, i));
    }
    if (t->len == 1) fputc(',', fp);
    fputc(')', fp);
}

static void py_format_dict(FILE *fp, PyObject *o) {
    PyDictObject *d = (PyDictObject *)o;
    fputc('{', fp);
    int64_t emitted = 0;
    for (int64_t i = 0; i < d->entries_used; i++) {
        DictEntry *e = &d->entries[i];
        if (e->key == NULL) continue;  /* dead slot from a prior delete */
        if (emitted > 0) fputs(", ", fp);
        py_format_repr(fp, pcc_gc_load_ptr(o, &e->key));
        fputs(": ", fp);
        py_format_repr(fp, pcc_gc_load_ptr(o, &e->value));
        emitted++;
    }
    fputc('}', fp);
}

static void py_format_set(FILE *fp, PyObject *o) {
    PySetObject *s = (PySetObject *)o;
    if (s->size == 0) {
        fputs("set()", fp);   /* empty set is set(), not {} */
        return;
    }
    fputc('{', fp);
    int64_t emitted = 0;
    for (int64_t i = 0; i < s->capacity; i++) {
        PyObject *key = s->entries[i].key;
        if (key == NULL || key == py_set_dummy) continue;  /* empty/tombstone */
        if (emitted > 0) fputs(", ", fp);
        py_format_repr(fp, pcc_gc_load_ptr(o, &s->entries[i].key));
        emitted++;
    }
    fputc('}', fp);
}

static void py_format_repr(FILE *fp, PyObject *o) {
    if (o == NULL) {
        fputs("<null>", fp);
        return;
    }
    if (PY_IS_TAGGED_INT(o)) {
        py_format_int(fp, o);
        return;
    }
    int32_t tag = py_header(o)->type_tag;
    if (tag == PY_TYPE_STR) {
        py_format_str_repr(fp, o);
        return;
    }
    if (tag == PY_TYPE_BYTES) {
        py_format_bytes(fp, o);
        return;
    }
    if (tag == PY_TYPE_BYTEARRAY) {
        py_format_bytearray(fp, o);
        return;
    }
    if (tag == PY_TYPE_INSTANCE || tag == PY_TYPE_EXC || tag >= PY_TYPE_USER) {
        /* repr() of a user instance must dispatch __repr__, not __str__.
         * Container elements (list/tuple/dict/set) recurse through here, so a
         * class with both __str__ and __repr__ would otherwise show __str__
         * inside a list. Falling through to py_format would call py_obj_str.
         * PY_TYPE_EXC: repr([KeyError('m')]) == [KeyError('m')], not the str.
         * On NULL (no __repr__) fall through to py_format's default handling. */
        PyObject *s = py_obj_repr(o);
        if (s != NULL) {
            py_format_str(fp, s);
            py_decref(s);
            return;
        }
    }
    py_format(fp, o);
}

static void py_format(FILE *fp, PyObject *o) {
    if (o == NULL) {
        fputs("<null>", fp);
        return;
    }
    if (PY_IS_TAGGED_INT(o)) {
        py_format_int(fp, o);
        return;
    }
    int32_t tag = py_header(o)->type_tag;
    switch (tag) {
        case PY_TYPE_NONE:
            fputs("None", fp);
            break;
        case PY_TYPE_BOOL:
            fputs(o == py_True ? "True" : "False", fp);
            break;
        case PY_TYPE_INT:
            py_format_int(fp, o);
            break;
        case PY_TYPE_FLOAT:
            py_format_float(fp, o);
            break;
        case PY_TYPE_STR:
            py_format_str(fp, o);
            break;
        case PY_TYPE_BYTES:
            py_format_bytes(fp, o);
            break;
        case PY_TYPE_BYTEARRAY:
            py_format_bytearray(fp, o);
            break;
        case PY_TYPE_LIST:
            py_format_list(fp, o);
            break;
        case PY_TYPE_TUPLE:
            py_format_tuple(fp, o);
            break;
        case PY_TYPE_DICT:
            py_format_dict(fp, o);
            break;
        case PY_TYPE_SET:
            py_format_set(fp, o);
            break;
        case PY_TYPE_COROUTINE:
            fputs("<coroutine object>", fp);
            break;
        case PY_TYPE_CONTINUATION:
            fputs("<continuation object>", fp);
            break;
        case PY_TYPE_VIRTUAL_THREAD:
            fputs("<virtual thread object>", fp);
            break;
        case PY_TYPE_EXC: {
            /* str(exc) == str of its single message value; py_exc_get_message
             * returns a borrowed ref (no decref). An arg-less exception (NULL
             * message) renders as the empty string. KeyError is special: its
             * __str__ is repr(key) (CPython str(KeyError('x'))=="'x'"). */
            PyObject *msg = py_exc_get_message(o);
            if (msg != NULL) {
                if (py_exc_matches(
                        o, (PyObject *)py_exc_builtin_class(PY_EXC_KEYERROR))) {
                    PyObject *r = py_obj_repr(msg);
                    if (r != NULL) {
                        py_format(fp, r);
                        py_decref(r);
                    }
                } else {
                    py_format(fp, msg);
                }
            }
            break;
        }
        default: {
            /* User-class instances and other objects: str(x) via py_obj_str
             * dispatches __str__ (then __repr__) — what print() uses in
             * CPython. Without it print(<instance>) rendered "<object tag=N>"
             * even when the class defines __str__. Safe from recursion: the
             * tags py_obj_str routes back through py_format_obj_to_str
             * (float/none/list/tuple/dict/set/bytes) are all handled above
             * this default, so the default only reaches the py_user_str_dispatch
             * path. NULL (no __str__/__repr__) falls back to the hook then
             * "<object tag=N>". */
            PyObject *s = py_obj_str(o);
            if (s != NULL) {
                py_format_str(fp, s);
                py_decref(s);
            } else {
                int fd = fileno(fp);
                if (fd < 0 || !py_format_try_cpy_object_into_fd(fd, o, tag)) {
                    fprintf(fp, "<object tag=%d>", (int)tag);
                }
            }
            break;
        }
    }
}

void py_print(PyObject *o) {
    py_format(stdout, o);
    fputc('\n', stdout);
}

/* Render any object to a freshly owned PyStr using the same formatting the
 * print path uses.  ``use_repr`` selects the repr form (quoted strings); for
 * containers/float str and repr coincide.  Used by ``py_obj_str`` /
 * ``py_obj_repr`` for the non-scalar types those paths do not handle inline.
 * Returns NULL on allocation failure (caller degrades gracefully). */
PyObject *py_format_obj_to_str(PyObject *o, int use_repr) {
    char *buf = NULL;
    size_t len = 0;
    FILE *ms = open_memstream(&buf, &len);
    if (ms == NULL) {
        return NULL;
    }
    if (use_repr) {
        py_format_repr(ms, o);
    } else {
        py_format(ms, o);
    }
    fclose(ms);
    PyObject *out = py_str_new(buf, (int64_t)len);
    free(buf);
    return out;
}

void py_print_many(PyObject *args_tuple, PyObject *sep, PyObject *end) {
    const char *sep_str = " ";
    size_t sep_len = 1;
    const char *end_str = "\n";
    size_t end_len = 1;

    if (sep != NULL && sep != py_None && py_type_of(sep) == PY_TYPE_STR) {
        PyStrObject *s = (PyStrObject *)sep;
        sep_str = s->data;
        sep_len = (size_t)s->byte_len;
    }
    if (end != NULL && end != py_None && py_type_of(end) == PY_TYPE_STR) {
        PyStrObject *s = (PyStrObject *)end;
        end_str = s->data;
        end_len = (size_t)s->byte_len;
    }

    if (args_tuple == NULL) {
        fwrite(end_str, 1, end_len, stdout);
        return;
    }

    PyTupleObject *t = (PyTupleObject *)args_tuple;
    for (int64_t i = 0; i < t->len; i++) {
        if (i > 0) fwrite(sep_str, 1, sep_len, stdout);
        py_format(stdout, fmt_tuple_item(args_tuple, t, i));
    }
    fwrite(end_str, 1, end_len, stdout);
}

File: pcc/py_runtime/py/py_print_fmt.py
"""pcc-Python port of py_print_fmt.c.

High-level print formatting for runtime PyObject* values. Native scalar
float printing is still emitted directly by codegen; this module covers
object-path print(), list/tuple repr, and print_many.
"""
from pcc.extern import extern, c_abi_export, c_ptr, c_int32, c_int64, c_void
from pcc.py_runtime.py.py_abi_constants import (
    PY_TYPE_CONTINUATION,
    PY_TYPE_VIRTUAL_THREAD,
    PY_TYPE_VTHREAD_CHANNEL,
)
from pcc.py_runtime.py.py_abi_constants import (
    PY_TYPE_BOOL,
    PY_TYPE_BYTEARRAY,
    PY_TYPE_BYTES,
    PY_TYPE_COROUTINE,
    PY_TYPE_DICT,
    PY_TYPE_EXC,
    PY_TYPE_FLOAT,
    PY_TYPE_INSTANCE,
    PY_TYPE_INT,
    PY_TYPE_LIST,
    PY_TYPE_NONE,
    PY_TYPE_SET,
    PY_TYPE_STR,
    PY_TYPE_TUPLE,
    PY_TYPE_USER_CLASS_START,
)
from pcc.unsafe import (
    cstr,
    free,
    global_load_ptr,
    is_tagged_int,
    load_i8,
    load_i32,
    load_i64,
    load_ptr,
    malloc,
    null,
    ptr_add,
    ptr_eq,
    ptr_is_null,
    store_i8,
    strlen,
    write,
)


py_int_value_i64 = extern("py_int_value_i64", (c_ptr,), c_int64)
py_bigint_to_cstr = extern("py_bigint_to_cstr", (c_ptr,), c_ptr)
py_float_format_fixed = extern("py_float_format_fixed", (c_ptr, c_int64), c_ptr)
py_float_repr_shortest = extern("py_float_repr_shortest", (c_ptr,), c_ptr)
pcc_gc_load_ptr = extern("pcc_gc_load_ptr", (c_ptr, c_ptr), c_ptr)
py_decref = extern("py_decref", (c_ptr,), c_void)
py_exc_get_message = extern("py_exc_get_message", (c_ptr,), c_ptr)
py_obj_str = extern("py_obj_str", (c_ptr,), c_ptr)
py_obj_repr = extern("py_obj_repr", (c_ptr,), c_ptr)
py_exc_matches = extern("py_exc_matches", (c_ptr, c_ptr), c_int64)
py_exc_builtin_class = extern("py_exc_builtin_class", (c_int64,), c_ptr)
fflush = extern("fflush", (c_ptr,), c_int32)
# Returns 1 if the unknown-tag object was a CPython PyObject and was
# rendered via PyObject_Str (libpython mode). Returns 0 in strict
# no-libpython mode (the hook variable is NULL there) so the caller
# falls through to ``<object tag=N>``. See py_process.c.
py_format_try_cpy_object_into_fd = extern(
    "py_format_try_cpy_object_into_fd",
    (c_int32, c_ptr, c_int32),
    c_int32,
)
# Render a numpy / C-extension scalar by driving its own tp_repr slot (NOT
# PyObject_Repr, which routes back to pcc and raises for a foreign object).
pcc_capi_cext_object_repr = extern("pcc_capi_cext_object_repr", (c_ptr,), c_ptr)
pcc_capi_is_cext_type_tag = extern("pcc_capi_is_cext_type_tag", (c_int64,), c_int64)


def _type_of(obj) -> int:
    if is_tagged_int(obj):
        return PY_TYPE_INT       # PY_TYPE_INT
    return load_i32(obj, 8)


def _write_cstr(p) -> None:
    if ptr_is_null(p) != 0:
        return
    n: int = strlen(p)
    if n > 0:
        write(1, p, n)


def _write_lit(p, n: int) -> None:
    write(1, p, n)


def _flush_stdio() -> None:
    # Keep ordering stable with codegen-emitted printf() calls for
    # native scalar print paths. fflush(NULL) flushes all output streams.
    fflush(null())


def _write_i64(v: int) -> None:
    if v == 0:
        _write_lit(cstr("0"), 1)
        return
    if v < 0:
        _write_lit(cstr("-"), 1)
        v = 0 - v
    buf = malloc(32)
    if ptr_is_null(buf) != 0:
        return
    n: int = 0
    while v > 0:
        digit: int = v % 10
        store_i8(buf, n, 48 + digit)
        n = n + 1
        v = v // 10
    i: int = n - 1
    while i >= 0:
        write(1, ptr_add(buf, i), 1)
        i = i - 1
    free(buf)


def _write_hex_nibble(v: int) -> None:
    ch: int = 0
    if v < 10:
        ch = 48 + v
    else:
        ch = 87 + v
    buf = malloc(1)
    if ptr_is_null(buf) != 0:
        return
    store_i8(buf, 0, ch)
    write(1, buf, 1)
    free(buf)


def _write_hex2(v: int) -> None:
    _write_hex_nibble((v >> 4) & 15)
    _write_hex_nibble(v & 15)


def _format_int(o) -> None:
    if is_tagged_int(o):
        _write_i64(py_int_value_i64(o))
        return
    s = py_bigint_to_cstr(o)
    if ptr_is_null(s) != 0:
        return
    _write_cstr(s)
    free(s)


def _format_str(o) -> None:
    n: int = load_i64(o, 16)
    if n > 0:
        write(1, ptr_add(o, 40), n)


def _format_str_repr(o) -> None:
    _write_lit(cstr("'"), 1)
    data = ptr_add(o, 40)
    n: int = load_i64(o, 16)
    i: int = 0
    while i < n:
        c: int = load_i8(data, i) & 255
        if c == 92:                  # '\\'
            _write_lit(cstr("\\\\"), 2)
        elif c == 39:                # "'"
            _write_lit(cstr("\\'"), 2)
        elif c == 10:
            _write_lit(cstr("\\n"), 2)
        elif c == 13:
            _write_lit(cstr("\\r"), 2)
        elif c == 9:
            _write_lit(cstr("\\t"), 2)
        elif c < 32 or c == 127:
            _write_lit(cstr("\\x"), 2)
            _write_hex2(c)
        else:
            write(1, ptr_add(data, i), 1)
        i = i + 1
    _write_lit(cstr("'"), 1)


def _format_bytes(o) -> None:
    _write_lit(cstr("b'"), 2)
    data = ptr_add(o, 24)
    n: int = load_i64(o, 16)
    i: int = 0
    while i < n:
        c: int = load_i8(data, i) & 255
        if c == 92:                  # '\\'
            _write_lit(cstr("\\\\"), 2)
        elif c == 39:                # "'"
            _write_lit(cstr("\\'"), 2)
        elif c == 10:
            _write_lit(cstr("\\n"), 2)
        elif c == 13:
            _write_lit(cstr("\\r"), 2)
        elif c == 9:
            _write_lit(cstr("\\t"), 2)
        elif c < 32 or c >= 127:
            # bytes repr: only printable ASCII (32..126) raw; control, DEL and
            # all high bytes (>=128) escape as \xNN. (c == 127 missed 128..255,
            # so b'\xcf\x80' printed the raw UTF-8 char instead of the escapes.)
            _write_lit(cstr("\\x"), 2)
            _write_hex2(c)
        else:
            write(1, ptr_add(data, i), 1)
        i = i + 1
    _write_lit(cstr("'"), 1)


def _format_bytearray(o) -> None:
    # bytearray repr: bytearray(b'...'); byte escaping is identical to bytes
    # (same layout: byte_len @16, data @24), wrapped in bytearray( ... ).
    _write_lit(cstr("bytearray("), 10)
    _format_bytes(o)
    _write_lit(cstr(")"), 1)


def _format_float(o) -> None:
    # CPython shortest-round-trip repr (handles inf/nan and the trailing ".0"
    # internally), via the shared C helper py_float_repr_shortest. The old path
    # used py_float_format_fixed(o, 6) which printed "3.333333" for 10/3.
    s = py_float_repr_shortest(o)
    if ptr_is_null(s) != 0:
        return
    data = ptr_add(s, 40)
    n: int = load_i64(s, 16)
    write(1, data, n)


def _format_list(o) -> None:
    _write_lit(cstr("["), 1)
    length: int = load_i64(o, 16)
    items = load_ptr(o, 32)
    i: int = 0
    while i < length:
        if i > 0:
            _write_lit(cstr(", "), 2)
        _format_repr(pcc_gc_load_ptr(o, ptr_add(items, i * 8)))
        i = i + 1
    _write_lit(cstr("]"), 1)


def _format_tuple(o) -> None:
    _write_lit(cstr("("), 1)
    length: int = load_i64(o, 16)
    i: int = 0
    while i < length:
        if i > 0:
            _write_lit(cstr(", "), 2)
        _format_repr(pcc_gc_load_ptr(o, ptr_add(o, 24 + i * 8)))
        i = i + 1
    if length == 1:
        _write_lit(cstr(","), 1)
    _write_lit(cstr(")"), 1)


def _format_dict(o) -> None:
    # PyDictObject: entries ptr @40, entries_used @48.
    # DictEntry (24 bytes): hash @0, key @8, value @16; key NULL = dead slot.
    _write_lit(cstr("{"), 1)
    entries = load_ptr(o, 40)
    entries_used: int = load_i64(o, 48)
    emitted: int = 0
    i: int = 0
    while i < entries_used:
        entry = ptr_add(entries, i * 24)
        if ptr_is_null(load_ptr(entry, 8)) == 0:
            if emitted > 0:
                _write_lit(cstr(", "), 2)
            _format_repr(pcc_gc_load_ptr(o, ptr_add(entry, 8)))
            _write_lit(cstr(": "), 2)
            _format_repr(pcc_gc_load_ptr(o, ptr_add(entry, 16)))
            emitted = emitted + 1
        i = i + 1
    _write_lit(cstr("}"), 1)


def _format_set(o) -> None:
    # PySetObject: size @16, capacity @24, entries ptr @40.
    # SetEntry (16 bytes): key @8 (NULL empty, py_set_dummy tombstone).
    size: int = load_i64(o, 16)
    if size == 0:
        _write_lit(cstr("set()"), 5)
        return
    _write_lit(cstr("{"), 1)
    entries = load_ptr(o, 40)
    dummy = global_load_ptr("py_set_dummy")
    cap: int = load_i64(o, 24)
    emitted: int = 0
    i: int = 0
    while i < cap:
        key = load_ptr(entries, i * 16 + 8)
        if ptr_is_null(key) == 0:
            if ptr_eq(key, dummy) == 0:
                if emitted > 0:
                    _write_lit(cstr(", "), 2)
                _format_repr(pcc_gc_load_ptr(o, ptr_add(entries, i * 16 + 8)))
                emitted = emitted + 1
        i = i + 1
    _write_lit(cstr("}"), 1)


def _format_repr(o) -> None:
    if ptr_is_null(o) != 0:
        _write_lit(cstr("<null>"), 6)
        return
    if is_tagged_int(o):
        _format_int(o)
        return
    tag: int = load_i32(o, 8)
    if tag == PY_TYPE_STR:                    # PY_TYPE_STR
        _format_str_repr(o)
        return
    if tag == PY_TYPE_BYTES:                   # PY_TYPE_BYTES
        _format_bytes(o)
        return
    if tag == PY_TYPE_BYTEARRAY:                   # PY_TYPE_BYTEARRAY
        _format_bytearray(o)
        return
    if (
        tag == PY_TYPE_INSTANCE
        or tag == PY_TYPE_EXC
        or tag >= PY_TYPE_USER_CLASS_START
    ):
        # repr() of a user instance must dispatch __repr__, not __str__.
        # Container elements recurse here, so a class with both __str__ and
        # __repr__ would otherwise show __str__ inside a list. PY_TYPE_EXC (12):
        # repr([KeyError('m')]) == [KeyError('m')], not the str. On NULL (no
        # __repr__) fall through to _format's default handling.
        s = py_obj_repr(o)
        if ptr_is_null(s) == 0:
            _format_str(s)
            py_decref(s)
            return
    _format(o)


def _format(o) -> None:
    if ptr_is_null(o) != 0:
        _write_lit(cstr("<null>"), 6)
        return
    if is_tagged_int(o):
        _format_int(o)
        return

    tag: int = load_i32(o, 8)
    if tag == PY_TYPE_NONE:                    # PY_TYPE_NONE
        _write_lit(cstr("None"), 4)
    elif tag == PY_TYPE_BOOL:                  # PY_TYPE_BOOL
        if ptr_eq(o, global_load_ptr("py_True")) != 0:
            _write_lit(cstr("True"), 4)
        else:
            _write_lit(cstr("False"), 5)
    elif tag == PY_TYPE_INT:                  # PY_TYPE_INT
        _format_int(o)
    elif tag == PY_TYPE_FLOAT:                  # PY_TYPE_FLOAT
        _format_float(o)
    elif tag == PY_TYPE_STR:                  # PY_TYPE_STR
        _format_str(o)
    elif tag == PY_TYPE_BYTES:                 # PY_TYPE_BYTES
        _format_bytes(o)
    elif tag == PY_TYPE_BYTEARRAY:                 # PY_TYPE_BYTEARRAY
        _format_bytearray(o)
    elif tag == PY_TYPE_LIST:                  # PY_TYPE_LIST
        _format_list(o)
    elif tag == PY_TYPE_TUPLE:                  # PY_TYPE_TUPLE
        _format_tuple(o)
    elif tag == PY_TYPE_DICT:                  # PY_TYPE_DICT
        _format_dict(o)
    elif tag == PY_TYPE_SET:                  # PY_TYPE_SET
        _format_set(o)
    elif tag == PY_TYPE_COROUTINE:                 # PY_TYPE_COROUTINE
        _write_lit(cstr("<coroutine object>"), 18)
    elif tag == PY_TYPE_CONTINUATION:
        _write_lit(cstr("<continuation object>"), 21)
    elif tag == PY_TYPE_VIRTUAL_THREAD:
        _write_lit(cstr("<virtual thread object>"), 23)
    elif tag == PY_TYPE_VTHREAD_CHANNEL:
        _write_lit(cstr("<vthread channel object>"), 24)
    elif tag == PY_TYPE_EXC:                 # PY_TYPE_EXC
        # str(exc) is the str of its single message value (CPython: the
        # exception args). py_exc_get_message returns a borrowed ref, so
        # no decref here; an arg-less exception (NULL message) renders as
        # the empty string. KeyError is special: its __str__ is repr(key)
        # (CPython str(KeyError('x'))=="'x'").
        msg = py_exc_get_message(o)
        if ptr_is_null(msg) == 0:
            if py_exc_matches(o, py_exc_builtin_class(4)) != 0:  # PY_EXC_KEYERROR
                r = py_obj_repr(msg)
                if ptr_is_null(r) == 0:
                    _format_str(r)
                    py_decref(r)
            else:
                _format(msg)
    else:
        # User-class instances and other objects: str(x) routes through
        # py_obj_str, which dispatches __str__ (then __repr__). This is what
        # print() uses in CPython; without it print(<instance>) rendered the
        # opaque "<object tag=N>" even when the class defines __str__. A NULL
        # result (no __str__/__repr__) falls back to the libpython hook then
        # "<object tag=N>".
        s = py_obj_str(o)
        if ptr_is_null(s) != 0:
            # numpy / C-extension scalar (e.g. an ndarray element): py_obj_str
            # has no pcc dispatch for it, so drive its own tp_repr slot rather
            # than printing the opaque <object tag=N>.
            if pcc_capi_is_cext_type_tag(tag) != 0:
                s = pcc_capi_cext_object_repr(o)
        if ptr_is_null(s) == 0:
            _format_str(s)
            py_decref(s)
        elif py_format_try_cpy_object_into_fd(1, o, tag) == 0:
            _write_lit(cstr("<object tag="), 12)
            _write_i64(tag)
            _write_lit(cstr(">"), 1)


@c_abi_export("py_print")
def py_print(o) -> None:
    _flush_stdio()
    _format(o)
    _write_lit(cstr("\n"), 1)


@c_abi_export("py_print_many")
def py_print_many(args_tuple, sep, end) -> None:
    _flush_stdio()
    sep_data = cstr(" ")
    sep_len: int = 1
    end_data = cstr("\n")
    end_len: int = 1
    none = global_load_ptr("py_None")

    if ptr_is_null(sep) == 0:
        if ptr_eq(sep, none) == 0:
            if _type_of(sep) == PY_TYPE_STR:
                sep_data = ptr_add(sep, 40)
                sep_len = load_i64(sep, 16)
    if ptr_is_null(end) == 0:
        if ptr_eq(end, none) == 0:
            if _type_of(end) == PY_TYPE_STR:
                end_data = ptr_add(end, 40)
                end_len = load_i64(end, 16)

    if ptr_is_null(args_tuple) != 0:
        write(1, end_data, end_len)
        return

    length: int = load_i64(args_tuple, 16)
    i: int = 0
    while i < length:
        if i > 0:
            write(1, sep_data, sep_len)
        _format(pcc_gc_load_ptr(args_tuple, ptr_add(args_tuple, 24 + i * 8)))
        i = i + 1
    write(1, end_data, end_len)
